src/activityLog/Activity.ts

```diff
@@ -0,0 +1,58 @@
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+
+import { AzExtParentTreeItem, AzExtTreeItem, IParsedError, parseError } from "@microsoft/vscode-azext-utils";
+import { TreeItemCollapsibleState } from "vscode";
+
+export interface ActivityTreeItemOptions {
+    label: string;
+    contextValuePostfix?: string;
+    collapsibleState?: TreeItemCollapsibleState;
+    children?: (parent: AzExtParentTreeItem) => AzExtTreeItem[];
+}
+
+interface ActivityType {
+    inital(): ActivityTreeItemOptions;
+    onSuccess(): ActivityTreeItemOptions;
+    onError(error: IParsedError): ActivityTreeItemOptions;
+}
+
+export type ActivityTask<R = void> = R extends void ? () => Promise<void> : () => Promise<R>;
+
+export abstract class ActivityBase<R = void> implements ActivityType {
+
+    abstract inital(): ActivityTreeItemOptions;
+    abstract onSuccess(): ActivityTreeItemOptions;
+    abstract onError(error: IParsedError): ActivityTreeItemOptions;
+
+    public done: boolean;
```

Comment:
This is one of those times I sure wish TypeScript would take the `foo { get; private set; }` from C#. Still, I'd make `done`, `error`, and `startedAtMs` private with a public `get` method. I think that means you can also mark them `readonly` on the interface, TS seems to treat `readonly` on interfaces as read-only to externals, whereas on classes as read-only to everyone (including the class itself).